tools/snitch/snitcher/src/lib.rs

```diff
@@ -390,38 +357,46 @@ pub fn raise_alert(
-        .to_url()
-        .and_then(|alert_url| {
-            HttpsConnector::new(4)
-                .map(|connector| (alert_url, connector))
-                .map_err(Error::from)
-        })
+    HttpsConnector::new(4)
+        .map(|connector| (settings.alert().url().clone(), connector))
+        .map_err(Error::from)
         .map(|(alert_url, connector)| {
-            let client = client::Client::new(
-                HyperClientService::new(HyperClient::builder().build(connector)),
-                alert_url,
-            );
-
-            Either::A(
-                client
-                    .request::<JsonValue, ()>(
-                        Method::POST,
-                        settings.alert().path(),
-                        Some(
-                            settings
-                                .alert()
-                                .query()
-                                .iter()
-                                .map(|(k, v)| (k.as_str(), v.as_str()))
-                                .collect(),
-                        ),
-                        Some(report_json),
-                        false,
-                    )
-                    .map(|_| ()),
-            )
+            let mut builder = Request::builder();
+            let uri = alert_url.as_str().parse::<Uri>().expect("Unexpected Url to Uri conversion failure");
+            let req = builder.method(Method::POST).uri(uri);
+            let serialized = serde_json::to_string(&report_json).unwrap();
+            req.header(CONTENT_TYPE, "text/json");
+            req.header(CONTENT_LENGTH, format!("{}", serialized.len()).as_str());
+
+            let hyper_client = HyperClient::builder().build(connector);
+            let request = req.body(Body::from(serialized)).unwrap();
+            debug!("send request to {}", request.uri());
+            let result = hyper_client
+                .request(request)
+                .map_err(move |err| {
+                    error!("HTTP request to {:?} failed with {:?}", alert_url, err);
+                    Error::from(err)
+                })
+                .and_then(|resp| {
+                    let status = resp.status();
+                    debug!("HTTP request succeeded with status {}", status);
+                    resp.into_body()
+                        .concat2()
+                        .map(move |body| (status, body))
+                        .map_err(|err| {
+                            error!("Reading response body failed with {:?}", err);
+                            Error::from(err)
+                        })
+                        .and_then(move |(status, body)| {
+                            if status.is_success() {
+                                Ok(())
+                            } else {
+                                Err(Error::from((status, &*body)))
+                            }
+                        })
+                });
+
+            Either::A(result.into_future())
```

Comment:
`result` is already a future. Shouldn't need the `into_future` call here.

Reply:
updated